# Melbourne Suburb Analytics

## Section 1: Data Cleaning

## 1.1 Setup

### 1.1.1 Configure Libraries

In [2]:
from pathlib import Path

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
pd.set_option('display.max_columns', None)

### 1.1.2 Load Dataset

In [4]:
proj_root = Path.cwd().parent
df_raw = pd.read_csv(proj_root/"data"/"raw"/"melb_data.csv")

df = df_raw.copy()

In [5]:
df.head()

,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,Postcode,Bedroom2,Bathroom,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount
0,Abbotsford,85 Turner St,2,h,1480000.0,S,Biggin,3/12/2016,2.5,3067.0,2.0,1.0,1.0,202.0,NaN,NaN,Yarra,-37.7996,144.9984,Northern Metropolitan,4019.0
1,Abbotsford,25 Bloomburg St,2,h,1035000.0,S,Biggin,4/02/2016,2.5,3067.0,2.0,1.0,0.0,156.0,79.0,1900.0,Yarra,-37.8079,144.9934,Northern Metropolitan,4019.0
2,Abbotsford,5 Charles St,3,h,1465000.0,SP,Biggin,4/03/2017,2.5,3067.0,3.0,2.0,0.0,134.0,150.0,1900.0,Yarra,-37.8093,144.9944,Northern Metropolitan,4019.0
3,Abbotsford,40 Federation La,3,h,850000.0,PI,Biggin,4/03/2017,2.5,3067.0,3.0,2.0,1.0,94.0,NaN,NaN,Yarra,-37.7969,144.9969,Northern Metropolitan,4019.0
4,Abbotsford,55a Park St,4,h,1600000.0,VB,Nelson,4/06/2016,2.5,3067.0,3.0,1.0,2.0,120.0,142.0,2014.0,Yarra,-37.8072,144.9941,Northern Metropolitan,4019.0


### 1.2 Initial Observations

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  str    
 1   Address        13580 non-null  str    
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  str    
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  str    
 6   SellerG        13580 non-null  str    
 7   Date           13580 non-null  str    
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  float64
 10  Bedroom2       13580 non-null  float64
 11  Bathroom       13580 non-null  float64
 12  Car            13518 non-null  float64
 13  Landsize       13580 non-null  float64
 14  BuildingArea   7130 non-null   float64
 15  YearBuilt      8205 non-null   float64
 16  CouncilArea    12211 non-null  str    
 17  Lattitude      13580 non-null  float64
 18  Longtitude     13

In [7]:
df["Car"].isna().sum()

np.int64(62)

In [8]:
df["BuildingArea"].isna().sum()

np.int64(6450)

In [9]:
df["YearBuilt"].isna().sum()

np.int64(5375)

In [10]:
df["CouncilArea"].isna().sum()

np.int64(1369)

In [11]:
df.drop(columns = ["Postcode"]).describe().style.format(lambda x: f"{x:6g}")

,Rooms,Price,Distance,Bedroom2,Bathroom,Car,Landsize,BuildingArea,YearBuilt,Lattitude,Longtitude,Propertycount
count,13580,13580,13580,13580,13580,13518,13580,7130,8205,13580,13580,13580
mean,2.938,1.07568e+06,10.1378,2.91473,1.53424,1.61008,558.416,151.968,1964.68,-37.8092,144.995,7454.42
std,0.955748,639311,5.86872,0.965921,0.691712,0.962634,3990.67,541.015,37.2738,0.0792598,0.103916,4378.58
min,1,85000,0,0,0,0,0,0,1196,-38.1825,144.432,249
25%,2,650000,6.1,2,1,1,177,93,1940,-37.8568,144.93,4380
50%,3,903000,9.2,3,1,2,440,126,1970,-37.8024,145,6555
75%,3,1.33e+06,13,3,2,2,651,174,1999,-37.7564,145.058,10331
max,10,9e+06,48.1,20,8,10,433014,44515,2018,-37.4085,145.526,21650


<sub>◦ *"Postcode" column is dropped since it is categorical data;*</sub>  
<sub>◦ *style.format() is used to remove unnecessary decimals* </sub>

In [12]:
df.duplicated().sum()

np.int64(0)

**Missing values by column**
- `Car` : 62 missing
- `BuildingArea` : 6,450 missing
- `YearBuilt` : 5,375 missing
- `CouncilArea` : 1,369 missing

**Datatype errors**
- `Date` : string
- `Postcode` : float (categorical)

**Unusual Numbers**
- `BuildingArea` and `Landsize` shows a min of 0 & an absurd max size
- `Price` has a very large max

**Duplicated rows**
- There are `0` duplicated rows

### 1.3 Data Types

In [13]:
df["Date"] = pd.to_datetime(df["Date"], format = "%d/%m/%Y")

In [14]:
df["Postcode"] = df["Postcode"].astype(str)

**Conversions**  
<sub>  Filtering & Computation  </sub>
- `Date` : string -> datetime 

<sub>  Categorical </sub>
- `Postcode` : float -> string  

### 1.4 Missing Values

In [15]:
df.fillna({"Car": df["Car"].median()}, inplace = True)

,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,Postcode,Bedroom2,Bathroom,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount
0,Abbotsford,85 Turner St,2,h,1480000.0,S,Biggin,2016-12-03,2.5,3067.0,2.0,1.0,1.0,202.0,NaN,NaN,Yarra,-37.79960,144.99840,Northern Metropolitan,4019.0
1,Abbotsford,25 Bloomburg St,2,h,1035000.0,S,Biggin,2016-02-04,2.5,3067.0,2.0,1.0,0.0,156.0,79.0,1900.0,Yarra,-37.80790,144.99340,Northern Metropolitan,4019.0
2,Abbotsford,5 Charles St,3,h,1465000.0,SP,Biggin,2017-03-04,2.5,3067.0,3.0,2.0,0.0,134.0,150.0,1900.0,Yarra,-37.80930,144.99440,Northern Metropolitan,4019.0
3,Abbotsford,40 Federation La,3,h,850000.0,PI,Biggin,2017-03-04,2.5,3067.0,3.0,2.0,1.0,94.0,NaN,NaN,Yarra,-37.79690,144.99690,Northern Metropolitan,4019.0
4,Abbotsford,55a Park St,4,h,1600000.0,VB,Nelson,2016-06-04,2.5,3067.0,3.0,1.0,2.0,120.0,142.0,2014.0,Yarra,-37.80720,144.99410,Northern Metropolitan,4019.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13575,Wheelers Hill,12 Strada Cr,4,h,1245000.0,S,Barry,2017-08-26,16.7,3150.0,4.0,2.0,2.0,652.0,NaN,1981.0,NaN,-37.90562,145.16761,South-Eastern Metropolitan,7392.0
13576,Williamstown,77 Merrett Dr,3,h,1031000.0,SP,Williams,2017-08-26,6.8,3016.0,3.0,2.0,2.0,333.0,133.0,1995.0,NaN,-37.85927,144.87904,Western Metropolitan,6380.0
13577,Williamstown,83 Power St,3,h,1170000.0,S,Raine,2017-08-26,6.8,3016.0,3.0,2.0,4.0,436.0,NaN,1997.0,NaN,-37.85274,144.88738,Western Metropolitan,6380.0
13578,Williamstown,96 Verdon St,4,h,2500000.0,PI,Sweeney,2017-08-26,6.8,3016.0,4.0,1.0,5.0,866.0,157.0,1920.0,NaN,-37.85908,144.89299,Western Metropolitan,6380.0


**Replacement of values**
- `Car` : Empty cells were replaced with `median()` of the `Car` column since there were only 62 missing cells, therefore being safe to fill.
- `BuildingArea`, `YearBuilt`, `CouncilArea` : Empty cells were not changed or replaced, left as `NaN` to avoid affecting calculations.

### 1.5 Outliers
#### 1.5.1 Observations

**`Price` (large value):**

In [45]:
thresholds = [0, 1000000, 2000000, 3000000, 4000000, 5000000, 6000000, 7000000, 8000000]

for price in thresholds:
    properties_over_threshold = (df["Price"]> price).sum()
    threshold_percent_base = properties_over_threshold/((df["Price"]>0).sum())*100
    print(f"Properties over ${price}: {properties_over_threshold}; {properties_over_threshold} to base ratio: {threshold_percent_base}%")

Properties over $0: 13580; 13580 to base ratio: 100.0%
Properties over $1000000: 5743; 5743 to base ratio: 42.29013254786451%
Properties over $2000000: 1007; 1007 to base ratio: 7.415316642120766%
Properties over $3000000: 221; 221 to base ratio: 1.6273932253313699%
Properties over $4000000: 55; 55 to base ratio: 0.40500736377025043%
Properties over $5000000: 21; 21 to base ratio: 0.15463917525773196%
Properties over $6000000: 6; 6 to base ratio: 0.044182621502209134%
Properties over $7000000: 3; 3 to base ratio: 0.022091310751104567%
Properties over $8000000: 1; 1 to base ratio: 0.007363770250368188%


In [17]:
df.sort_values("Price", ascending= False).head(21)[["Suburb", "Price", "Type", "Landsize", "BuildingArea", "Regionname"]]

,Suburb,Price,Type,Landsize,BuildingArea,Regionname
12094,Mulgrave,9000000.0,h,744.0,117.0,South-Eastern Metropolitan
7692,Canterbury,8000000.0,h,2079.0,464.3,Southern Metropolitan
9575,Hawthorn,7650000.0,h,1690.0,284.0,Southern Metropolitan
3616,Kew,6500000.0,h,1334.0,365.0,Southern Metropolitan
12557,Middle Park,6400000.0,h,553.0,308.0,Southern Metropolitan
6372,Toorak,6250000.0,h,564.0,342.0,Southern Metropolitan
7554,Brighton,5800000.0,h,1276.0,NaN,Southern Metropolitan
5631,South Yarra,5700000.0,h,292.0,272.0,Southern Metropolitan
9233,Middle Park,5600000.0,h,472.0,328.0,Southern Metropolitan
251,Armadale,5525000.0,h,1491.0,516.0,Southern Metropolitan


- ~7.415% of properties are over `$2M`
    - Crosses the 10% mark
    - ~92.585% of all houses are under `$2M`
- ~1.627% of properties are over `$3M`
    - Drops from 1007 to 221 houses, massive reduction
- ~0.405% of properties are over `$4M`
    - Less than 0.5% of properties
    - Huge outlier
- ~0.1546% of properties are over `$5M`
    - 21 properties exist
    - In the top 21, property `Type` are all houses `h` which makes sense since only sale price of houses be over `$5M` compared to an apartment or a townhouse
- ~0.007364% of properties are over `$8M`
    - 1 property exist
    - Property is worth the most despite having the smallest `Landsize` and `BuildingArea` out of the top 4, which is suprising it would be more expected for a top valued property to have a much larger `Landsize`
    - Only property above `$5M` that is located in South-Eastern Metropolitan rather than Southern Metropolitan from `Regionname`, hinting that this property may be extremely unique, considering its smaller `Landsize`
    - Meaning this property is probable to exist

**`Landsize` min == 0**:

In [18]:
pd.crosstab(df["Landsize"] == 0, df["Type"])

Type,h,t,u
Landsize,,,
False,9278,975,1388
True,171,139,1629


- Mostly exist in units `u` (1629), much greater than houses `h` (171) or townhouses `t` (139)

In [19]:
pd.crosstab(df["Landsize"] == 0, df["Type"],  normalize="columns")

Type,h,t,u
Landsize,,,
False,0.981903,0.875224,0.46006
True,0.018097,0.124776,0.53994


- Normalizing by column shows ~54% of units have a `Landsize` of 0 compared to ~1.8% for houses and ~12.5% for townhouses
    - This happens since units/aparentments often share land with a building so a 0 could be a "not applicable" filler since the entire floor could be owned and measured collectively
        - Hence I will be using the 0 values in my data analysis since it is arguably correct (no individual land rather than missing value)
    - For houses, the 2% rate means more likely it could ben an error or missing value
        - For houses and townhouses, by definition they cannot have 0m² of land.
        - Therefore for `Landsize` == 0 where `Type` are `h` or `t`, `Landsize` will be set to **`NaN`**

In [20]:
df.loc[(df["Type"]=="u") & (df["Landsize"]>0), "Landsize"].describe()

count     1388.000000
mean      1037.505043
std       2432.107456
min          1.000000
25%        166.000000
50%        292.000000
75%        982.250000
max      37000.000000
Name: Landsize, dtype: float64

- For units `u` with a landsize above 0, the distribution is right-skewed since median is ~1037, min is 1 and max is 37000
    - This gap means a small number of extreme outliers pulling the average upwards 
    - 37000m² is definitely not possible for a single unit, so it might be a recording of a whole complex/building rather than a single unit, which means there might be inconsistencies of what is classified and measured for `u`

**`BuildingArea` min = 0**:

In [21]:
pd.crosstab(df["BuildingArea"] == 0, df["Type"])

Type,h,t,u
BuildingArea,,,
False,9435,1111,3017
True,14,3,0


- Opposite of Landsize, houses `h` have 14 rows, townhouses `t` have 3 and units `u` have 0

In [22]:
pd.crosstab(df["BuildingArea"] == 0, df["Type"], normalize="columns")

Type,h,t,u
BuildingArea,,,
False,0.998518,0.997307,1.0
True,0.001482,0.002693,0.0


- Reflected in the normalizing by column, with all properties with `BuildingArea` 0 being under 0.003 for its `Type`
    - All units having a `BuildingArea` makes sense since it is not possible to have an apartment with 0 internal floor space.
    - Since `h` and `t` have such small numbers, the 0's are most likely missing values that were not looked at during data entry
    - Therefore for `BuildingArea` == 0, `BuildingArea` will be set to **`NaN`**

**`Landsize` large max value**:

In [23]:
landsize_top = df.loc[:, ["Suburb", "Type", "Price", "Landsize", "BuildingArea", "Regionname"]].sort_values("Landsize", ascending= False).head(20).reset_index(drop=False)
landsize_top.index+=1
landsize_top

,index,Suburb,Type,Price,Landsize,BuildingArea,Regionname
1,11020,Fitzroy,h,2700000.0,433014.0,NaN,Northern Metropolitan
2,10504,Silvan,h,1085000.0,76000.0,NaN,Eastern Victoria
3,687,Balwyn North,h,2000000.0,75100.0,NaN,Southern Metropolitan
4,13245,New Gisborne,h,1355000.0,44500.0,44515.0,Northern Victoria
5,5194,Reservoir,h,572000.0,41400.0,NaN,Northern Metropolitan
6,11371,Gisborne,h,807000.0,40468.0,NaN,Northern Victoria
7,10045,Silvan,h,1070000.0,38490.0,NaN,Eastern Victoria
8,9223,Maribyrnong,u,585000.0,37000.0,90.0,Western Metropolitan
9,3942,Maribyrnong,u,495000.0,37000.0,NaN,Western Metropolitan
10,8241,Port Melbourne,u,1030000.0,21715.0,99.0,Southern Metropolitan


- 2nd, 4th, 6th, 7th highest `Landsize` houses `h` have `Regionname` outside metropolitan areas
    - Probable since their locations are in sem-rural hill areas
- 3rd, 5th highest `Landsize` houses `h` have `Regionname` inside metropolitan areas
    - In metropolian areas, there are strict rules in place for extremely large sizes of private homes.
        - The largest recorded Balwyn North house (mega mansion) is ~1000 square meters
        - The largest recorded Reservoir house is ~500 square meters
    - If those were its real recorded `Landsize` the property's `Price` will be much higher than $2M or $500K
    - Hence for `Landsize` = 75100 and 41400 (index = 687, 5194), `Landsize` will be set to **`NaN`**

In [24]:
print(f"Top 2 difference: +{landsize_top.loc[1, "Landsize"]-landsize_top.loc[2, "Landsize"]}, x{landsize_top.loc[1, "Landsize"]/landsize_top.loc[2, "Landsize"]}")

Top 2 difference: +357014.0, x5.697552631578947


In [25]:
print(f"99th percentile: {df["Landsize"].quantile(0.99)}")
print(f"Ratio to 1st item: x{df["Landsize"].max()/df["Landsize"].quantile(0.99)}")

99th percentile: 2959.82999999998
Ratio to 1st item: x146.29691570124058


- 1st (top) `Landsize` is 357014m² more than the 2nd highest `Landsize` 
    - This is 5 times the 2nd highest `Landsize`, which is 76000m²
    - It's area is ~146x the 99th percentile (2959.83)
    - Located in Fitzroy = ~1,000,000m² (accounting streets and boundaries), which a house `h` taking up 1/2 in one of the most densely populated suburbs is physically impossible at all
    - Hence for `Landsize` = 433014 (index = 11020), `Landsize` will be set to **`NaN`**

In [26]:
## standard outlier rule: Q3 + 1.5*IQR, where IQR is Q3-Q1
landsize_Q1 = df["Landsize"].quantile(0.25)
landsize_Q3 = df["Landsize"].quantile(0.75)
landsize_IQR = landsize_Q3-landsize_Q1
landsize_upper_outlier = landsize_Q3+(1.5*landsize_IQR)

print(f"Upper Outlier: {landsize_upper_outlier}")
print(f"Amount of outliers w/Type 'u': {((df["Type"]=="u") & (df["Landsize"]>landsize_upper_outlier)).sum()}")
df.loc[(df["Type"]=="u") & (df["Landsize"]>landsize_upper_outlier),["Suburb", "Type", "Price", "Landsize", "BuildingArea", "Regionname"]].sort_values("Landsize", ascending= False)

Upper Outlier: 1362.0
Amount of outliers w/Type 'u': 248


,Suburb,Type,Price,Landsize,BuildingArea,Regionname
3942,Maribyrnong,u,495000.0,37000.0,NaN,Western Metropolitan
9223,Maribyrnong,u,585000.0,37000.0,90.00,Western Metropolitan
8241,Port Melbourne,u,1030000.0,21715.0,99.00,Southern Metropolitan
4706,Port Melbourne,u,1340000.0,21700.0,NaN,Southern Metropolitan
10819,South Yarra,u,762500.0,17200.0,79.00,Southern Metropolitan
...,...,...,...,...,...,...
7602,Brunswick,u,315000.0,1403.0,NaN,Northern Metropolitan
9177,Hawthorn,u,690000.0,1396.0,95.88,Southern Metropolitan
11302,Brunswick West,u,275000.0,1388.0,36.00,Northern Metropolitan
9976,Murrumbeena,u,357000.0,1382.0,NaN,Southern Metropolitan


In [27]:
print(f"Rows with duplicated landsize: {df.loc[(df["Type"]=="u") & (df["Landsize"]>landsize_upper_outlier),"Landsize"].duplicated(keep = False).sum()}")
print(f"Percent of duplicated outliers: {(df.loc[(df["Type"]=="u") & (df["Landsize"]>landsize_upper_outlier),"Landsize"].duplicated(keep = False).sum())/(((df["Type"]=="u") & (df["Landsize"]>landsize_upper_outlier)).sum())*100}%")

Rows with duplicated landsize: 70
Percent of duplicated outliers: 28.225806451612907%


- For units `u`, there were 248 properties with `Landsize` > 1362m² (upper outlier: Q3+1.5*IQR)
    - There are 70 duplicated units `u` that are above the upper outlier, or ~28% of the outliers
        - Duplicated values most likely mean they belong to the **same development** site
        - Meaning they are likely to represent the total land area of the development rather than individual units
        - Hence it is safer to keep the data as original rather than claiming an error

**`BuildingArea` large max value:**

In [28]:
building_area_top = df.loc[:,["Suburb", "Type", "Price", "Landsize", "BuildingArea", "Regionname"]].sort_values("BuildingArea", ascending= False).head(20).reset_index(drop=False)
building_area_top.index+=1
building_area_top

,index,Suburb,Type,Price,Landsize,BuildingArea,Regionname
1,13245,New Gisborne,h,1355000.0,44500.0,44515.0000,Northern Victoria
2,1484,Bulleen,h,1280000.0,732.0,6791.0000,Eastern Metropolitan
3,2560,Fitzroy North,t,930000.0,2778.0,3558.0000,Northern Metropolitan
4,1588,Camberwell,h,2608000.0,730.0,3112.0000,Southern Metropolitan
5,2234,Elsternwick,u,741000.0,0.0,1561.0000,Southern Metropolitan
6,2830,Glen Iris,t,600000.0,0.0,1143.0000,Southern Metropolitan
7,1368,Brunswick,h,1053000.0,343.0,1041.0000,Northern Metropolitan
8,3640,Kew,h,1975000.0,531.0,1022.0000,Southern Metropolitan
9,12064,Lalor,h,580500.0,532.0,934.0000,Northern Metropolitan
10,10920,Brighton,h,3000000.0,818.0,826.8367,Southern Metropolitan


In [47]:
print(f"Top 2 difference: +{building_area_top.loc[1, "BuildingArea"]-building_area_top.loc[2, "BuildingArea"]}, x{building_area_top.loc[1, "BuildingArea"]/building_area_top.loc[2, "BuildingArea"]}")
print(f"2nd V 3rd difference: +{building_area_top.loc[2, "BuildingArea"]-building_area_top.loc[3, "BuildingArea"]}, x{building_area_top.loc[2, "BuildingArea"]/building_area_top.loc[3, "BuildingArea"]}")

print(f"99th percentile: {df["BuildingArea"].quantile(0.99)}")
print(f"99th percentile to 1st item: x{df["BuildingArea"].max()/df["BuildingArea"].quantile(0.99)}")

Top 2 difference: +37724.0, x6.55499926373141
2nd V 3rd difference: +3233.0, x1.9086565486228217
99th percentile: 463.47600000000074
99th percentile to 1st item: x3.36802768643899


- The top `BuildingArea` located in New Gisborne is 37724m² more than the 2nd highest `BuildingArea` 
    - This is 6.5 times the 2nd highest `BuildingArea`, which is 466m²
    - For comparison, the difference between 2nd and 3rd highest `BuildingArea` is only 3233m²
    - It's area is ~95x the 99th percentile (2959.83)

In [48]:
print(f"99.95th percentile: {df["BuildingArea"].quantile(0.9995)}")
print(f"4th and 5th highest 'BuildingArea' difference: +{building_area_top.loc[4,"BuildingArea"]-building_area_top.loc[5,"BuildingArea"]}, x{building_area_top.loc[4,"BuildingArea"]/building_area_top.loc[5,"BuildingArea"]}")

99.95th percentile: 973.2480000000724
4th and 5th highest 'BuildingArea' difference: +1551.0, x1.9935938500960924


- Top 4 observations (3000+) lay far from even the 99.95th percentile (extreme upper tail)
    - 4th highest `BuildingArea` is double (~x1.99) the size of 5th highest `BuildingArea` with a difference of 1551
    - Because of their absurd values compared to the rest of the data, for `BuildingArea` = 44515, 6791, 3558 and 3112 (index = 13245, 1484, 2560, 1588), `BuildingArea` will be set to **`NaN`**

In [33]:
## standard outlier rule: Q3 + 1.5*IQR, where IQR is Q3-Q1
building_area_Q1 = df["BuildingArea"].quantile(0.25)
building_area_Q3 = df["BuildingArea"].quantile(0.75)
building_area_IQR = building_area_Q3-building_area_Q1
building_area_Q3+(1.5*building_area_IQR)

np.float64(295.5)

- Upper outlier is plausible (295.5), can very much exist for multistory buildings.
    - Therefore not used for extreme outlier reasoning

#### 1.5.2 Masking

For `Landsize` == 0 where `Type` are `h` or `t`, `Landsize` -> **`NaN`**

In [34]:
df.loc[(df["Landsize"]==0) & (df["Type"].isin(["h", "t"])), "Landsize"] = np.nan

For `BuildingArea` == 0, `BuildingArea` -> **`NaN`**

In [35]:
df.loc[df["BuildingArea"]==0, "BuildingArea"] = np.nan

For `Landsize` = 433014, `Landsize` -> **`NaN`**

In [36]:
df.loc[11020, "Landsize"] = np.nan

For `Landsize` = 75100 and 41400, `Landsize` -> **`NaN`**

In [37]:
df.loc[[687, 5194], "Landsize"] = np.nan

For `BuildingArea` = 44515, 6791, 3558 and 3112, `BuildingArea` -> **`NaN`**

In [38]:
df.loc[[13245, 1484, 2560, 1588], "BuildingArea"] = np.nan

### 1.6 Validation

In [39]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Suburb         13580 non-null  str           
 1   Address        13580 non-null  str           
 2   Rooms          13580 non-null  int64         
 3   Type           13580 non-null  str           
 4   Price          13580 non-null  float64       
 5   Method         13580 non-null  str           
 6   SellerG        13580 non-null  str           
 7   Date           13580 non-null  datetime64[us]
 8   Distance       13580 non-null  float64       
 9   Postcode       13580 non-null  str           
 10  Bedroom2       13580 non-null  float64       
 11  Bathroom       13580 non-null  float64       
 12  Car            13580 non-null  float64       
 13  Landsize       13267 non-null  float64       
 14  BuildingArea   7109 non-null   float64       
 15  YearBuilt      8205 non-null  

In [40]:
checks = {
    "Landsize==0 (h|t)": df.loc[(df["Landsize"]==0) & (df["Type"].isin(["h", "t"])), "Landsize"],
    "BuildingArea==0": df.loc[df["BuildingArea"]==0, "BuildingArea"],
    "Landsize large outliers": df.loc[[687, 5194, 11020], "Landsize"],
    "BuildingArea large outliers": df.loc[[13245, 1484, 2560, 1588], "BuildingArea"],
}

for key, v in checks.items():
    print(f"{key} all NaN: {v.isna().all()}")

Landsize==0 (h|t) all NaN: True
BuildingArea==0 all NaN: True
Landsize large outliers all NaN: True
BuildingArea large outliers all NaN: True


### 1.7 Export

In [41]:
df.to_csv(proj_root/"data"/"processed"/"melb_data_cleaned.csv", index = False)